src/main/java/dev/dfonline/codeclient/dev/overlay/ActionViewer.java

```diff
@@ -0,0 +1,184 @@
+package dev.dfonline.codeclient.dev.overlay;
+
+import com.google.common.collect.Lists;
+import dev.dfonline.codeclient.CodeClient;
+import dev.dfonline.codeclient.config.Config;
+import dev.dfonline.codeclient.dev.menu.customchest.CustomChestMenu;
+import dev.dfonline.codeclient.hypercube.actiondump.Action;
+import dev.dfonline.codeclient.hypercube.actiondump.ActionDump;
+import dev.dfonline.codeclient.location.Dev;
+import net.minecraft.block.entity.ChestBlockEntity;
+import net.minecraft.block.entity.SignBlockEntity;
+import net.minecraft.client.gui.DrawContext;
+import net.minecraft.client.gui.screen.ingame.GenericContainerScreen;
+import net.minecraft.client.gui.screen.ingame.HandledScreen;
+import net.minecraft.client.gui.tooltip.TooltipBackgroundRenderer;
+import net.minecraft.client.gui.tooltip.TooltipComponent;
+import net.minecraft.client.gui.tooltip.TooltipPositioner;
+import net.minecraft.client.item.TooltipContext;
+import net.minecraft.item.ItemStack;
+import net.minecraft.item.WrittenBookItem;
+import net.minecraft.nbt.NbtCompound;
+import net.minecraft.text.Style;
+import net.minecraft.text.Text;
+import net.minecraft.util.Formatting;
+import net.minecraft.util.hit.BlockHitResult;
+import net.minecraft.util.thread.ThreadExecutor;
+import org.joml.Vector2i;
+import org.joml.Vector2ic;
+
+import java.util.Iterator;
+import java.util.List;
+
+public class ActionViewer {
+    private static Action action;
+    private static boolean reference = false;
+    private static ItemStack book;
+
+    public static void invalidate() {
+        action = null;
+        reference = false;
+        book = null;
+    }
+    public static void onClickChest(BlockHitResult hitResult) {
+        invalidate();
+        var world = CodeClient.MC.world;
+        if (world == null || !Config.getConfig().ActionViewer) return;
+        var position = hitResult.getBlockPos();
+        if (CodeClient.location instanceof Dev dev && dev.isInDev(position) && world.getBlockEntity(position) instanceof ChestBlockEntity) {
+            var signEntity = world.getBlockEntity(position.down().west());
+            if (signEntity instanceof SignBlockEntity sign) {
+                if (sign.getFrontText().getMessage(0, true).contains(Text.literal("CALL FUNCTION"))) {
+                    reference = true;
+                    return;
+                }
+                try {
+                    action = ActionDump.getActionDump().findAction(sign.getFrontText());
+                } catch (Exception ignored) {
+
+                }
+            }
+        }
+    }
+
+    public static ItemStack getReferenceBook() {
```

Comment:
This should be a utility, not just for this feature. We should also probably construct a hypercube.actiondump.Action from the reference book, which would allow usage of the ref book in the parameters ghosts.
You can just move it outside (to Utility perhaps) but you don't need to construct an action from it

Reply:
converting the reference book to an action was going to be my first approach but i hadn't wanted mess with parsing all of that data. however, i have submitted a commit that adds a new class `hypercube.ReferenceBook` that can be used for that purpose in the future with parameters ghosts and also moved `ActionViewer#getReferenceBook()` into `location.Dev` as it felt fitting to go there.